### ISSUE TITLE
[Feature Request]: Less strict numeric conversions writing to BigQuery with Storage API

### ISSUE THREAD

**Issue by bastewart** [apache/beam#25228]
### What would you like to happen?

Currently a number of type conversions are very strict in [`TableRowToStorageApiProto`](https://github.com/apache/beam/blob/634b0453469b66ee4c135aca48b02d2425916f36/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java) leading to potentially unnecessary loss of data.

As an example the `INT64` conversion will fail if the value arriving is a Java Double, even if that double could be cast to an integer will no loss of data/precision. Below the double `36183.0` fails to convert to an `INT64` even though a more optimistic conversion would succeed:

```
Exception: org.apache.beam.sdk.io.gcp.bigquery.TableRowToStorageApiProto$SchemaDoesntMatchException: Unexpected value :36183.0, type: class java.lang.Double. Table field name: <snip>, type: INT64'}
```

(The colon is misleading, it was a typo [that's been fixed in a newer commit](https://github.com/apache/beam/commit/b896cd5ab6ddd993ae2819ce5f40f43bab707459).)

We could have used `Number.longValue` (or `Math.round`) here and verified there was no loss of precision by checking equality afterwards.

Code here for this example: https://github.com/apache/beam/blob/634b0453469b66ee4c135aca48b02d2425916f36/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java#L691-L712

In the current implementation I don't think NUMERIC and BIGNUMERIC do not accept `BigInteger` at the moment either. I think my issue and this would be solved with a fall-through which optimistically converts all `Number`s to a long and checking for loss of precision afterwards.

In general this coupled with failing rows with any invalid values (#25227) means it's quite easy for a lot of data to fail to write.

I'd be happy to open a PR here, but have never contributed here and Java isn't my primary language, so it may be a little slow...

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

**Comment by reuvenlax** [comment]
This is an interesting request. Even in the above example, not all INT64 values can be converted to Java double. e.g. the following code:

long longValue = XXXX;
double doubleValue = (double) longValue;
assert(longValue == (long) doubleValue);

will often fail, especially for large numbers (e.g. try it out for 499999999000000001L).

I guess we could convert optimistically and fail if the specific number happens to lose precision, however I'm uncomfortable with that. That's a recipe for a pipeline that works for a while and then suddenly starts behaving poorly as the input data changes, which is very confusing. I think understandable and easily-predictable behavior is better.

**Comment by reuvenlax** [comment]
It also should be easy for any user to write a ParDo (or MapElements) that does their own numeric conversions before calling BigQueryIO.

**Comment by bastewart** [comment]
Sorry, I had missed you reply!

Summary on this message is I think in our case it's very hard to do the conversion. I think it's also reasonable to treat `1.0` as an integer, JSON schema [does for example](https://json-schema.org/understanding-json-schema/reference/numeric.html#integer). I also think it might also be a regression over the Streaming API implementation (at least this with #25227 and #25233 means there is a regression).

> Even in the above example, not all INT64 values can be converted to Java double. e.g. the following code:

Sorry, I may be misunderstanding, but that's not not the way round we have the issue.

Our issue is Java doubles arriving and needing to be converted to an `INT64` for BigQuery. In that case - by my understanding ([see my fix commit](https://github.com/bastewart/beam/commit/38601213f81896444c60dd9e590f8a795358d09a)) - it's trivial to optimistically  convert it to a long and fail if you've lost precision.

> It also should be easy for any user to write a ParDo (or MapElements) that does their own numeric conversions before calling BigQueryIO.

Unfortunately in our case this is very-much non-trivial 😅

We write to 1000+ tables which have dynamic schemas*. This means we're relying on Beam to load the schema and convert the values for us. I think we'd have to re-implement, or use directly, most of `StorageApiDynamicDestinations` to grab schema info, and then do our own traversal of all data before handing it over to Beam.

*I am aware that automatic schema reloading in the Beam Storage Writes API is not yet released...

We also don't have direct/easy control over the types are the arrive. We're consuming a JSON blob off Kafka and bundling the data into BigQuery. Those JSON blobs _have_ been validated against the BigQuery schema using [JSON Schema](https://json-schema.org/). JSON schema [validates numbers with `0` fractional part  as an integer](https://json-schema.org/understanding-json-schema/reference/numeric.html#integer), so this would be hard for us to guard against.

More generally, and personally, I feel like `1.0` should be treated as an integer. It's extremely easy for that kind of mis-conversion to occur, and being overly strict just leads to unexpected consequences. This issue coupled with #25227 and #25233 means that a lot of rows can be dropped "silently".

### Streaming API Behaviour

I need to check (will get back to you later!), but I am reasonably sure the Streaming API allows this type of coercion to occur. I'll double check and get back to you though.

**Comment by bastewart** [comment]
Yeah, I have experimented and the [Streaming API](https://cloud.google.com/bigquery/docs/reference/rest/v2/tabledata/insertAll) is much more lenient.

In general there can be extraneous (and "missing") decimal portions for all the numeric types. These are all accepted in the Streaming API for all of `INTEGER` (`INT64`), `NUMERIC`, `BIGNUMERIC` and `FLOAT` (`FLOAT64`):
 - `"1.0"`
 - `1.0`
 - `"1"`
 - `1`
 
Any non-zero decimal portion is not valid for an `INTEGER` though. So these are valid only for everything else:
 - `"1.1"`
 - `1.1`

This does mean that in addition to [my fix for double/float -> long](https://github.com/bastewart/beam/commit/38601213f81896444c60dd9e590f8a795358d09a) the current handling of `String` when an `INT64` is desired will fail for something like `"1.0"`, where the Streaming API would accept it: https://github.com/apache/beam/blob/dcfadddf52ae68a8d3d605e3a072c5d611b1fe14/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java#L693-L695
